### ISSUE TITLE
Systematic fusion batches

### ISSUE THREAD

**Issue by RolandBERTINJOHANNET** [ruflab/shimmer#43]
**systematic fusion manifesto**

### It makes more sense to use all the available matched data in each batch, instead of having random binary scores

So far in training fusion, we were fusing only once per batch with selection factors that could be 0.

_This meant we weren't using some of the matched samples._

For two modalities this gives us a 50% chance of using the information from either modality in each batch (on binary score cases)
But with 10 modalities, it goes down to 10%. This means that to use all the matched data you'd probably have to loop over the data around 30 times or so (so that the random sampling puts a one on each modality).

Doing **thirty times more epochs** and relying on stochasticity to decide when we use the data in the batches is strange.

So now we're going to implement systematic enacting of all the use cases inside the broadcast loss. That is, **exhaustively fuse with all different possible permutations of zeros and ones on all the incoming modalities.**

### First step : picking modalities for the broadcast loss

We first begin with encoding all the available latents.
Then, we get all the possible permutations of zeros and ones for available modalities (ex : [1,0,0], [1,1,0], [1,1,1],[0,1,0],[0,1,1],[0,0,1].......)

### Second step : linear scaling robustness

We want the global workspace to be capable of handling _any kind of linear scaling_ on the prefusion representations.
All the while, these scaling factors  must always sum to one.

So basically what we want is to transform scaling factors that look like this from _step 1_ : **[0,1,0,0,1,0,0]** (length==the number of modalities)
to something like this :  **[0,0.7,0,0,0.3,0,0]** 

This doesn't need to be done _explicitly_. It can be done while passing the non-zero-scaled latents to a **RandomAttention** module (which will just return uniform-then-softmaxxed scores), which will give us as many scaling factors as there are non-zero modalities, and then we can just apply them as usual, and then fuse.

_Note : for the case where we have a single one in the permutation, the exact same code will work as the softmax will assign a one to that modality._

### Third step : broadcast loss

At this point we have as many fused vectors as there are permutations between the available matched modalities.

We decode all of them back to all the representations for which we had data (even the ones we zeroed out), and compute the mse (compute_loss) on all of them.
Given that we're doing all permutations, this includes all possible translations, and on top of that all possible demi-cycles.

### Fourth step : usupervised losses

What happens for the unsupervised batch samples ?

The loss above, if we give it a unimodal sample, should work without an error, and give us a demi-cycle loss.

But since full cycles proved useful in the fusion-less global workspace, we'd like to add them in too.

Can we elegantly do cycles inside the broadcast loss in the same way ?

One possibility is this : 

```
nb_steps = 1 if len(latents.keys())>1 else 2

#encode representations     
for step_num in range(nb_steps):
    #get all permutations
    #apply linear scaling
    #fuse with these attention_scores
    #decode to original domains
    #compare with ground truth domain latents.
    ##the process should stop here if we're in the matched case
    #here, set representations to be the decoded representations
```

**But there are a few issues :** 

1. We don't want to re-do all the permutations in the second step of the loop
2. We'd like to just have one fusion when encoding for a second time, where we give random linear scaling to all the predicted domains, except the one where we have ground truth, which we set to 0 (to make it a one-to-many-to-one cycle).

**Why do we even want to do cycles inside the broadcast loss ?**
It'd be nice to just have two losses: the alignment loss, and the broadcast loss. It's much more intuitive, elegant and easy to introduce to people.

### Recap : why this is good : 

This version of the global workspace includes in its broadcast loss : 

1. the translation loss
2. the cycle loss
3. the demi-cycle loss
4. all the versions of the broadcast loss we previously tested and validated

In addition : 

1. It is directly scalable to more than two modalities
2. It takes full advantage of all matched data in each batch
3. Since it only encodes once, the added computational cost is only to decode once for each permutation.
